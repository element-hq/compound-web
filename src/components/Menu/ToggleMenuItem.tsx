/*
Copyright 2023, 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, {
  type ComponentProps,
  forwardRef,
  useCallback,
  useId,
} from "react";
import { MenuItem } from "./MenuItem";
import { ToggleInput } from "../Form/Controls/Toggle";

type Props = Pick<
  ComponentProps<typeof MenuItem>,
  "className" | "Icon" | "label" | "onSelect" | "disabled" | "onClick"
> & {
  /**
   * Whether the toggle is checked.
   */
  checked: boolean;
};

/**
 * A menu item with a toggle control.
 * Must be used within a compound Menu or other `menu` or `menubar` aria role subtree.
 */
export const ToggleMenuItem = forwardRef<HTMLInputElement, Props>(
  function ToggleMenuItem(
    { className, Icon, label, onSelect, checked, disabled, onClick },
    ref,
  ) {
    const toggleId = useId();
    // The toggle is controlled and we intend to ignore its events. We do need
    // to at least set onChange though to make React happy.
    const onChange = useCallback(() => {}, []);

    // <label> elements are not allowed to have a role like menuitemcheckbox, so
    // we must instead use a plain <div> for the menu item and use aria-checked
    // etc. to communicate its state.
    return (
      <MenuItem
        as="button"
        role="menuitemcheckbox"
        aria-checked={checked}
        className={className}
        Icon={Icon}
        label={label}
        hideChevron
        onSelect={onSelect}
        disabled={disabled}
        onClick={onClick}
      >
        <ToggleInput
          id={toggleId}
          ref={ref}
          // This is purely cosmetic; really the whole MenuItem is the toggle.
          aria-hidden
          tabIndex={-1}
          checked={checked}
          disabled={disabled}
          onChange={onChange}
        />
      </MenuItem>
    );
  },
);
