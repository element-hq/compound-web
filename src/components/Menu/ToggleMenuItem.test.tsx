/*
Copyright 2023, 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import React from "react";
import VisibilityOnIcon from "@vector-im/compound-design-tokens/assets/web/icons/visibility-on";

import { ToggleMenuItem } from "./ToggleMenuItem";
import userEvent from "@testing-library/user-event";

describe("ToggleMenuItem", () => {
  it("renders", () => {
    const { asFragment } = render(
      <ToggleMenuItem
        Icon={VisibilityOnIcon}
        label="Always show"
        checked={false}
        onSelect={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("toggles", async () => {
    const user = userEvent.setup();
    const toggle = vi.fn();
    render(
      <ToggleMenuItem
        Icon={VisibilityOnIcon}
        label="Always show"
        checked={false}
        onSelect={toggle}
      />,
    );

    // Try toggling using keyboard controls
    await user.tab();
    await user.keyboard("[Space]");
    expect(toggle).toBeCalledTimes(1);
    toggle.mockClear();

    // Try toggling by clicking
    await user.click(screen.getByRole("menuitemcheckbox"));
    expect(toggle).toBeCalledTimes(1);
    toggle.mockClear();

    // Also try clicking the label specifically (as this has regressed in the
    // past)
    await user.click(screen.getByText("Always show"));
    expect(toggle).toBeCalledTimes(1);
  });
});
