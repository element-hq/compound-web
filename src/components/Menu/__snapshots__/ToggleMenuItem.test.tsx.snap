// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleMenuItem > renders 1`] = `
<DocumentFragment>
  <div
    aria-checked="false"
    class="_item_9618f6 _interactive_9618f6"
    data-kind="primary"
    role="menuitemcheckbox"
  >
    <svg
      aria-hidden="true"
      class="_icon_9618f6"
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 16q1.875 0 3.188-1.312Q16.5 13.375 16.5 11.5t-1.312-3.187T12 7 8.813 8.313 7.5 11.5t1.313 3.188T12 16m0-1.8q-1.125 0-1.912-.787A2.6 2.6 0 0 1 9.3 11.5q0-1.125.787-1.912A2.6 2.6 0 0 1 12 8.8q1.125 0 1.912.787.788.788.788 1.913t-.787 1.912A2.6 2.6 0 0 1 12 14.2m0 4.8q-3.475 0-6.35-1.837Q2.775 15.324 1.3 12.2a.8.8 0 0 1-.1-.312 3 3 0 0 1 0-.775.8.8 0 0 1 .1-.313q1.475-3.125 4.35-4.962Q8.525 4 12 4t6.35 1.838T22.7 10.8a.8.8 0 0 1 .1.313 3 3 0 0 1 0 .774.8.8 0 0 1-.1.313q-1.475 3.125-4.35 4.963Q15.475 19 12 19m0-2a9.54 9.54 0 0 0 5.188-1.488A9.77 9.77 0 0 0 20.8 11.5a9.77 9.77 0 0 0-3.613-4.012A9.54 9.54 0 0 0 12 6a9.55 9.55 0 0 0-5.187 1.487A9.77 9.77 0 0 0 3.2 11.5a9.77 9.77 0 0 0 3.613 4.012A9.54 9.54 0 0 0 12 17"
      />
    </svg>
    <span
      class="_typography_61e060 _font-body-md-medium_61e060 _label_9618f6"
    >
      Always show
    </span>
    <div
      class="_container_d9763f"
    >
      <input
        aria-hidden="true"
        class="_input_d9763f"
        id="_r_0_"
        role="switch"
        type="checkbox"
      />
      <div
        class="_ui_d9763f"
      />
    </div>
  </div>
</DocumentFragment>
`;
